internal/blobcache/blobcache.go

```diff
@@ -0,0 +1,359 @@
+		Digest:       dgst.String(),
+		Size:         written,
+		LastAccessed: now,
+		CreatedAt:    now,
+		Sources: []Source{{
+			Registry:   registry,
+			Repository: repository,
+		}},
+	}
+
+	log.Debugf(ctx, "Cached blob %s (%d bytes) for %s/%s", dgst, written, registry, repository)
+
+	return bc.saveMetadata()
+}
+
+// blobPath returns the internal path for a blob.
+func (bc *BlobCache) blobPath(dgst digest.Digest) string {
+	return filepath.Join(bc.rootDir, "blobs", dgst.Algorithm().String(), dgst.Encoded())
+}
+
+func (bc *BlobCache) addSourceLocked(ctx context.Context, dgst digest.Digest, registry, repository string) error {
+	digestStr := dgst.String()
+
+	info, exists := bc.metadata.Blobs[digestStr]
+	if !exists {
+		// Blob file exists but metadata doesn't - reconstruct it.
+		blobPath := bc.blobPath(dgst)
+
+		stat, err := os.Stat(blobPath)
+		if err != nil {
+			log.Warnf(ctx, "Failed to stat blob %s at %s: %v", dgst, blobPath, err)
+
+			return fmt.Errorf("stat blob %s: %w", dgst, err)
+		}
+
+		info = BlobInfo{
+			Digest:       digestStr,
+			Size:         stat.Size(),
+			CreatedAt:    stat.ModTime(),
+			LastAccessed: time.Now(),
+		}
+	}
+
+	// Update last accessed time.
+	info.LastAccessed = time.Now()
+
+	// Check if source already exists.
+	for _, s := range info.Sources {
+		if s.Registry == registry && s.Repository == repository {
+			bc.metadata.Blobs[digestStr] = info
+
+			return bc.saveMetadata()
+		}
+	}
+
+	info.Sources = append(info.Sources, Source{Registry: registry, Repository: repository})
+	bc.metadata.Blobs[digestStr] = info
+
+	log.Debugf(ctx, "Added source %s/%s to blob %s", registry, repository, dgst)
+
+	return bc.saveMetadata()
+}
+
+// saveMetadata persists the metadata to disk. Caller must hold bc.mu.
+func (bc *BlobCache) saveMetadata() error {
+	data, err := json.MarshalIndent(bc.metadata, "", "  ")
+	if err != nil {
+		return fmt.Errorf("marshaling metadata: %w", err)
+	}
+
+	// Atomic write.
+	tmpFile, err := os.CreateTemp(filepath.Dir(bc.metadataPath), ".metadata-*.tmp")
+	if err != nil {
+		return fmt.Errorf("creating temp metadata: %w", err)
+	}
+
+	tmpPath := tmpFile.Name()
+	success := false
+
+	defer func() {
+		if !success {
+			os.Remove(tmpPath)
+		}
+	}()
+
+	if _, err := tmpFile.Write(data); err != nil {
+		tmpFile.Close()
+
+		return fmt.Errorf("writing metadata: %w", err)
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return fmt.Errorf("closing temp metadata: %w", err)
+	}
+
+	if err := os.Rename(tmpPath, bc.metadataPath); err != nil {
+		return err
+	}
+
+	success = true
+
+	return nil
+}
+
+// GCStats contains statistics from a garbage collection operation.
+type GCStats struct {
+	BlobsRemoved int
+	BytesFreed   int64
+}
+
+// GarbageCollect removes blobs that are no longer referenced by any image.
+// It returns statistics about the GC operation.
+func (bc *BlobCache) GarbageCollect(ctx context.Context, referencedDigests map[string]bool) (GCStats, error) {
```

Comment:
Do you know in which case there's orphaned image cache?
It seems the cache is tied to the image lifecycle. I wonder if we really need this Garbage collection mechanism instead of just a delete function.